packages/lesswrong/lib/sql/FragmentLexer.ts

```diff
@@ -0,0 +1,60 @@
+/**
+ * Util class for lexing GraphQL fragments.
+ */
+class FragmentLexer {
+  private name: string;
+  private baseTypeName: string;
+  private lines: string[];
+  private position = 0;
+
+  constructor(fragmentSrc: string) {
+    const match = fragmentSrc.match(/\s*fragment\s+([a-zA-Z0-9-_]+)\s+on\s+([a-zA-Z0-9-_]+)\s*{(.*)}/s);
+    if (match?.length !== 4) {
+      throw new Error("Cannot parse fragment text");
+    }
+
+    this.name = match[1];
+    this.baseTypeName = match[2];
+    this.lines = match[3].split("\n").map((s) => s.trim()).filter((s) => s);
+  }
+
+  /**
+   * The name of this fragment (eg; PostsMinimumInfo).
+   */
+  getName() {
+    return this.name;
+  }
+
+  /**
+   * The GraphQL type that this fragment is defined on (eg; Post).
+   */
+  getBaseTypeName() {
+    return this.baseTypeName;
+  }
+
+  /**
+   * Get the next line from the input. This is guaranteed to be a non-empty
+   * string with all whitespace and comments trimmed from the beginning and end.
+   * Returns null when we reach the end of the input text.
+   */
+  next(): string | null {
+    while (this.position < this.lines.length) {
+      const line = this.lines[this.position++];
+      const trimmedLine = line.match(/\s*([^#]+)(#.*)?/)?.[1]?.trimEnd();
```

Comment:
I'm going to leave this as-is for now as it's working, but I think long-term it might be better to actually generate the fields directly from the Apollo AST rather than adding an extra library (assuming we don't replace Apollo at some point). This would also allow us to support non-fragment graphiql queries (albeit, with worse performance as we couldn't cache the results).

RE the problem with comments, if anything is actually broken by this logic then it should already be broken because we apply pretty much the same logic in `registerFragment`.